akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -263,6 +364,232 @@ abstract class GraphStageLogic {
+   * signal (before invoking the `andThen` function).
+   */
+  final protected def emit[T](out: Outlet[T], elem: T, andThen: () ⇒ Unit): Unit =
+    if (isAvailable(out)) {
+      push(out, elem)
+      andThen()
+    } else {
+      setOrAddEmitting(out, new EmittingSingle(out, elem, getHandler(out), andThen))
+    }
+
+  /**
+   * Emit an element through the given outlet, suspending execution if necessary.
+   * This action replaces the [[OutHandler]] for the given outlet if suspension
+   * is needed and reinstalls the current handler upon receiving an `onPull()`
+   * signal.
+   */
+  final protected def emit[T](out: Outlet[T], elem: T): Unit = emit(out, elem, DoNothing)
+
+  /**
+   * Abort outstanding (suspended) emissions for the given outlet, if there are any.
+   * This will reinstall the replaced handler that was in effect before the `emit`
+   * call.
+   */
+  final protected def abortEmitting(out: Outlet[_]): Unit =
+    getHandler(out) match {
+      case e: Emitting[_] ⇒
+        setHandler(out, e.previous)
+      case _ ⇒
+    }
+
+  private def setOrAddEmitting[T](out: Outlet[T], next: Emitting[T]): Unit =
+    getHandler(out) match {
+      case e: Emitting[_] ⇒ e.asInstanceOf[Emitting[T]].addFollowUp(next)
+      case _              ⇒ setHandler(out, next)
+    }
+
+  private abstract class Emitting[T](protected val out: Outlet[T], val previous: OutHandler, andThen: () ⇒ Unit) extends OutHandler {
+    private var followUps: mutable.Queue[Emitting[T]] = null
+
+    protected def followUp(): Unit = {
+      setHandler(out, previous)
+      andThen()
+      if (followUps != null) {
+        val next = followUps.dequeue()
+        if (followUps.nonEmpty) next.followUps = followUps
+        setHandler(out, next)
+      }
+    }
+
+    def addFollowUp(e: Emitting[T]): Unit =
+      if (followUps == null) followUps = mutable.Queue(e)
+      else followUps.enqueue(e)
+
+    override def onDownstreamFinish(): Unit = previous.onDownstreamFinish()
+  }
+
+  private class EmittingSingle[T](_out: Outlet[T], elem: T, _previous: OutHandler, _andThen: () ⇒ Unit)
+    extends Emitting(_out, _previous, _andThen) {
+
+    override def onPull(): Unit = {
+      push(out, elem)
+      followUp()
+    }
+  }
+
+  private class EmittingIterator[T](_out: Outlet[T], elems: Iterator[T], _previous: OutHandler, _andThen: () ⇒ Unit)
+    extends Emitting(_out, _previous, _andThen) {
+
+    override def onPull(): Unit = {
+      push(out, elems.next())
+      if (!elems.hasNext) {
+        followUp()
+      }
+    }
+  }
+
+  /**
+   * Install a handler on the given inlet that emits received elements on the
+   * given outlet before pulling for more data. `doTerminate` controls whether
+   * completion or failure of the given inlet shall lead to stage termination or not.
+   */
+  final protected def passAlong[Out, In <: Out](from: Inlet[In], to: Outlet[Out], doTerminate: Boolean): Unit =
```

Comment:
my guess is that more dynamic usage will be better off just implementing the whole handler directly